erts/emulator/beam/jit/x86/instr_bs.cpp

```diff
@@ -1596,8 +1630,187 @@ struct BscSegment {
+            offsetof(struct erl_bits_state, erts_current_bin_);
+    const int cur_bin_offset =
+            offsetof(ErtsSchedulerRegisters, aux_regs.d.erl_bits_state) +
+            offsetof(struct erl_bits_state, erts_bin_offset_);
+
+    x86::Mem mem_bin_base =
+            x86::Mem(registers, cur_bin_base - x_reg_offset, sizeof(UWord));
+    x86::Mem mem_bin_offset =
+            x86::Mem(registers, cur_bin_offset - x_reg_offset, sizeof(UWord));
+
+    a.mov(bin_base, mem_bin_base);
+    a.mov(bin_offset, mem_bin_offset);
+
+    a.mov(ARG1, bin_offset);
+    if (size_reg.isValid()) {
+        a.add(mem_bin_offset, size_reg);
+    } else {
+        a.add(mem_bin_offset, imm(size));
+    }
+    a.shr(ARG1, imm(3));
+    a.add(ARG1, bin_base);
+}
+
+bool BeamModuleAssembler::need_mask(const ArgVal Val, Sint size) {
+    if (size == 64) {
+        return false;
+    } else if (always_small(Val)) {
+        auto [min, max] = getIntRange(Val);
+        return !(0 <= min && max >> size == 0);
+    } else {
+        return true;
+    }
+}
+
+void BeamModuleAssembler::set_zero(Sint effectiveSize) {
+    Label words = a.newLabel();
+    Label less_than_a_word = a.newLabel();
+    Sint store_unit = 1;
+
+    update_bin_state(RET, ARG2, -1, ARG3);
+
+    if (effectiveSize >= 128) {
+        /* Store 128-bit words when the size is known and at least 128
+         * bits. */
+        store_unit = 2;
+        a.xorps(x86::xmm0, x86::xmm0);
+    }
+
+    if (effectiveSize % 128 != 0) {
+        /* The size is either not known or 16-byte aligned. */
+        mov_imm(RET, 0);
+    }
+
+    if (effectiveSize < Sint(store_unit * 8 * sizeof(Eterm))) {
+        /* The size is either not known or smaller than a machine word. */
+        a.cmp(ARG3, imm(store_unit * 8 * sizeof(Eterm)));
+        a.short_().jl(less_than_a_word);
+    }
+
+    a.bind(words);
+    if (store_unit == 2) {
+        a.movups(x86::Mem(ARG1, 0, 16), x86::xmm0);
+    } else {
+        a.mov(x86::Mem(ARG1, 0, 8), RET);
+    }
+    a.add(ARG1, imm(store_unit * sizeof(Eterm)));
+    a.sub(ARG3, imm(store_unit * 8 * sizeof(Eterm)));
+    a.cmp(ARG3, imm(store_unit * 8 * sizeof(Eterm)));
+    a.jge(words);
+
+    a.bind(less_than_a_word);
+    if (effectiveSize < 0) {
+        /* Unknown size. */
+        Label byte_loop = a.newLabel();
+        Label done = a.newLabel();
+
+        ASSERT(store_unit = 1);
+
+        a.test(ARG3, ARG3);
+        a.short_().jz(done);
+
+        a.bind(byte_loop);
+        a.mov(x86::Mem(ARG1, 0, 1), RET.r8());
+        a.inc(ARG1);
+        a.sub(ARG3, imm(8));
+        a.short_().jg(byte_loop);
+
+        a.bind(done);
+    } else if (effectiveSize % (store_unit * 8 * sizeof(Eterm)) != 0) {
+        /* The size is known, and we know that there are less than
+         * 128 bits to initialize. */
+        ASSERT(store_unit = 1);
+
+        if ((effectiveSize & 127) >= 64) {
+            a.mov(x86::Mem(ARG1, 0, 8), RET);
+            a.add(ARG1, imm(8));
```

Comment:
It would be faster and more compact to use static offsets here and below, doing a single update of `ARG1` at the end if the caller expects it to be updated.

Reply:
Done.